### PR TITLE
[JENKINS-33971] ReverseBuildTrigger.upstream2Trigger cache fix

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-33971](https://issues.jenkins-ci.org/browse/JENKINS-33971)

@reviewbybees esp. @fbelzunc

### CODE REVIEW

File: core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java [hunk 1]
```diff
@@ -221,9 +199,53 @@ public FormValidation doCheckUpstreamProjects(@AncestorInPath Job project, @Quer
     }
 
     @Extension public static final class RunListenerImpl extends RunListener<Run> {
+        
+        static RunListenerImpl get() {
+            return ExtensionList.lookup(RunListener.class).get(RunListenerImpl.class);
```
Comment (amuniz) [inline core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java]: Why not `ExtensionList.lookup(RunListenerImpl.class)`?

**Comment by amuniz** [comment]
:bee:

File: core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java [hunk 2]
```diff
@@ -221,9 +199,53 @@ public FormValidation doCheckUpstreamProjects(@AncestorInPath Job project, @Quer
...
+
+        private Map<Job,Collection<ReverseBuildTrigger>> upstream2Trigger;
+
+        synchronized void invalidateCache() {
+            upstream2Trigger = null;
+        }
+
+        private Map<Job,Collection<ReverseBuildTrigger>> calculateCache() {
+            Map<Job,Collection<ReverseBuildTrigger>> result = new WeakHashMap<>();
+            SecurityContext orig = ACL.impersonate(ACL.SYSTEM);
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java]: Are you sure that <code>finally</code> handler is placed correctly? It's inside for loop :ant:

**Comment by oleg-nenashev** [comment]
Maybe the logic impersonation is correct, but I don't understand it in such case.
:bug: , because the second <code>for (Job<?, ?> downstream : Jenkins.getInstance().getAllItems(Job.class)) </code> iteration will be executed with the non-admin context

File: core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java]: Not sure if that even works—the extension point is `RunListener`. Either way you would need to have the call to `get` or similar anyway.

**Comment by oleg-nenashev** [comment]
:bee:

**Pull request merged**